explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java

```diff
@@ -0,0 +1,141 @@
+     * @see "Glover, Fred. "Tabu search—part I." ORSA Journal on computing 1, no. 3 (1989): 190-206"
+     */
+    public CounterfactualExplainer(DataBoundaries dataBoundaries, List<Boolean> contraints, List<Output> goal, Long timeLimit, int tabuSize, int acceptedCount) {
+        this(dataBoundaries, contraints, goal, CounterfactualConfigurationFactory.createSolverConfig(timeLimit, tabuSize, acceptedCount));
+    }
+
+    private List<CounterfactualEntity> createEntities(PredictionInput predictionInput) {
+        final List<CounterfactualEntity> entities = new ArrayList<>();
+
+        for (int i = 0; i < predictionInput.getFeatures().size(); i++) {
+            final Feature feature = predictionInput.getFeatures().get(i);
+            final Boolean isConstrained = constraints.get(i);
+            final FeatureBoundary featureDistribution = dataBoundaries.getFeatureBoundaries().get(i);
+            if (feature.getType() == Type.NUMBER) {
+                if (feature.getValue().getUnderlyingObject() instanceof Double) {
+                    final DoubleEntity doubleEntity = new DoubleEntity(feature, featureDistribution.getStart(), featureDistribution.getEnd(), isConstrained);
+                    entities.add(doubleEntity);
+                } else if (feature.getValue().getUnderlyingObject() instanceof Integer) {
+                    final IntegerEntity integerEntity = new IntegerEntity(feature, (int) featureDistribution.getStart(), (int) featureDistribution.getEnd(), isConstrained);
+                    entities.add(integerEntity);
+                }
+            } else if (feature.getType() == Type.BOOLEAN) {
+                final BooleanEntity booleanEntity = new BooleanEntity(feature, isConstrained);
+                entities.add(booleanEntity);
+            }
+        }
+        return entities;
+    }
+
+    @Override
+    public CompletableFuture<List<CounterfactualEntity>> explainAsync(Prediction prediction, PredictionProvider model) {
+
+        final List<CounterfactualEntity> entities = createEntities(prediction.getInput());
+
+        final UUID problemId = UUID.randomUUID();
+
+        SolverManager<CounterfactualSolution, UUID> solverManager =
+                SolverManager.create(solverConfig, new SolverManagerConfig());
+
+        CounterfactualSolution problem =
+                new CounterfactualSolution(entities, model, goal);
+
+
+        SolverJob<CounterfactualSolution, UUID> solverJob = solverManager.solve(problemId, problem);
+        CounterfactualSolution solution;
+        try {
+            // Wait until the solving ends
+            solution = solverJob.getFinalBestSolution();
+        } catch (InterruptedException | ExecutionException e) {
+            throw new IllegalStateException("Solving failed: {}", e);
+        }
+
+        solverManager.close();
+
+        return CompletableFuture.completedFuture(solution.getEntities());
```

Comment:
I think we can wrap the logic and make it async like
```suggestion
        return CompletableFuture.supplyAsync(() -> {
            SolverJob<CounterfactualSolution, UUID> solverJob = solverManager.solve(problemId, problem);
            CounterfactualSolution solution;
            try {
                // Wait until the solving ends
                solution = solverJob.getFinalBestSolution();
                return solution.getEntities();
            } catch (InterruptedException | ExecutionException e) {
                throw new IllegalStateException("Solving failed: {}", e);
            } finally {
                solverManager.close();
            }
        }, getExecutor());
```

Note: add an `Executor` field + constructor with `Executor`. Then the method `getExecutor()` should return that value or `ForkJoinPool.commonPool()` if no `Executor` has been explicitly provided

Reply:
Implemented